File: Data Structures & Algorithms/house-robber-ii/submission-2.cpp
class Solution {
public:
    int rob(vector<int>& nums) {

        if (nums.size() == 1)
        {
            return nums[0];
        }

        if (nums.size() == 2)
        {
            return std::max(nums[0], nums[1]);
        }

        int l = nums.size() - 1;
        vector<int> max_amount_1(l,0);
        vector<int> max_amount_2(l,0);

        max_amount_1[0] = nums[0];
        max_amount_1[1] = std::max(nums[0],nums[1]);

        max_amount_2[0] = nums[1];
        max_amount_2[1] = std::max(nums[1],nums[2]);

        for (int i=2; i<l; i++)
        {
            max_amount_1[i] = std::max(max_amount_1[i-2]+ nums[i], max_amount_1[i-1]);
            max_amount_2[i] = std::max(max_amount_2[i-2]+ nums[i+1], max_amount_2[i-1]);
        }
        
        return std::max(max_amount_1[l-1],max_amount_2[l-1]);
    }
};

File: Data Structures & Algorithms/house-robber/submission-2.py
class Solution:
    def rob(self, nums: List[int]) -> int:

        if len(nums)== 1:
            return nums[0]
        max_amount_robbed = [-1]*len(nums)
        max_amount_robbed[0] = nums[0]
        max_amount_robbed[1] = max(nums[0],nums[1])
        for i in range(2,len(nums)):
            max_amount_robbed[i] = max(max_amount_robbed[i-2] + nums[i], max_amount_robbed[i-1])
        
        return max_amount_robbed[-1]